opt: panic in NullRejectAggVar

```
I180819 04:21:55.784564 276 sql/conn_executor.go:1052  [n1,client=[::1]:49345,user=root] [Open pos:602] executing ExecStmt: SELECT (SELECT nspname FROM pg_catalog.pg_namespace LIMIT 1 OFFSET 4) AS c0 FROM crdb_internal.ranges AS ref_0 LEFT JOIN pg_catalog.pg_inherits AS ref_1 RIGHT JOIN pg_catalog.pg_user AS ref_2 ON (CAST(NULL AS TIMESTAMP WITH TIME ZONE) < CAST(NULL AS TIMESTAMP WITH TIME ZONE)) ON (EXISTS (SELECT ref_1.inhseqno AS c0 FROM information_schema.referential_constraints AS ref_3 WHERE (SELECT descriptor_ids FROM crdb_internal.jobs LIMIT 1 OFFSET 5) = ref_0.replicas LIMIT 118)) INNER JOIN crdb_internal.forward_dependencies AS ref_4 ON ((CAST(NULL AS TIMESTAMP[]) != CAST(NULL AS TIMESTAMP[])) AND (((CAST(NULL AS DATE) >= CAST(NULL AS TIMESTAMP WITH TIME ZONE)) OR (ref_4.dependedonby_id IS NULL)) OR (CAST(NULL AS DATE) != CAST(NULL AS DATE)))) WHERE CAST(COALESCE(ref_4.dependedonby_id, pg_catalog.lastval()) AS INT8) < ref_0.lease_holder
E180819 04:21:55.785712 276 sql/conn_executor.go:661  [n1,client=[::1]:49345,user=root] a SQL panic has occurred while executing "SELECT (SELECT nspname FROM pg_catalog.pg_namespace LIMIT 1 OFFSET 4) AS c0 FROM crdb_internal.ranges AS ref_0 LEFT JOIN pg_catalog.pg_inherits AS ref_1 RIGHT JOIN pg_catalog.pg_user AS ref_2 ON (CAST(NULL AS TIMESTAMP WITH TIME ZONE) < CAST(NULL AS TIMESTAMP WITH TIME ZONE)) ON (EXISTS (SELECT ref_1.inhseqno AS c0 FROM information_schema.referential_constraints AS ref_3 WHERE (SELECT descriptor_ids FROM crdb_internal.jobs LIMIT 1 OFFSET 5) = ref_0.replicas LIMIT 118)) INNER JOIN crdb_internal.f [...]": couldn't find an aggregate that is not ConstAgg/ConstNotNullAgg
I180819 04:21:55.785796 276 sql/conn_executor.go:681  [n1,client=[::1]:49345,user=root] finishing connExecutor
E180819 04:21:55.786275 276 util/log/crash_reporting.go:203  [n1,client=[::1]:49345,user=root] a panic has occurred!
E180819 04:21:55.790614 276 util/log/crash_reporting.go:477  [n1,client=[::1]:49345,user=root] Reported as error d14e159b7c884b1ea5817c120e7aaf34
panic: couldn't find an aggregate that is not ConstAgg/ConstNotNullAgg [recovered]
        panic: couldn't find an aggregate that is not ConstAgg/ConstNotNullAgg [recovered]
        panic: panic while executing 1 statements: SELECT (SELECT _ FROM _._ LIMIT _ OFFSET _) AS _ FROM _._ AS _ LEFT JOIN _._ AS _ RIGHT JOIN _._ AS _ ON (CAST(_ AS TIMESTAMP WITH TIME ZONE) < CAST(_ AS TIMESTAMP WITH TIME ZONE)) ON (EXISTS (SELECT _._ AS _ FROM _._ AS _ WHERE (SELECT _ FROM _._ LIMIT _ OFFSET _) = _._ LIMIT _)) INNER JOIN _._ AS _ ON ((CAST(_ AS TIMESTAMP[]) != CAST(_ AS TIMESTAMP[])) AND (((CAST(_ AS DATE) >= CAST(_ AS TIMESTAMP WITH TIME ZONE)) OR (_._ IS _)) OR (CAST(_ AS DATE) != CAST(_ AS DATE)))) WHERE CAST(COALESCE(_._ [...]; caused by couldn't find an aggregate that is not ConstAgg/ConstNotNullAgg
```

```
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build.func1(0xc4251aa998)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:130 +0x99
panic(0x5f53920, 0x688e4a0)
        /usr/local/Cellar/go/1.10.2/libexec/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).NullRejectAggVar(0xc425308f88, 0x4b0000004a, 0x4000000000000)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/reject_nulls.go:59 +0x146
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructSelect(0xc425308f78, 0x4f0000004b, 0x1400000000004f)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:881 +0x1859
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructSelect(0xc425308f78, 0x4e0000004c, 0x800000000004e)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:848 +0x1b0c
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructLeftJoinApply(0xc425308f78, 0x4c00000001, 0xc40000004e, 0xc4285fab40)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:4095 +0xb19
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).ConstructApplyJoin(0xc425308f88, 0x100000008, 0x4e0000004c, 0x1)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/decorrelate.go:260 +0xb9
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).HoistJoinSubquery(0xc425308f88, 0x100000008, 0x1f00000008, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/decorrelate.go:147 +0xa9
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructLeftJoin(0xc425308f78, 0x800000001, 0x1f, 0x1)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:2179 +0xd15
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).constructJoin(0xc427cb5960, 0x100000001, 0x1e00000008, 0xc4286320e8)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/join.go:348 +0xc5
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildJoin(0xc427cb5960, 0xc428628040, 0xc4285ffc00, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/join.go:72 +0x5cb
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildDataSource(0xc427cb5960, 0x68a3140, 0xc428628040, 0x0, 0xc4285ffc00, 0xc420b165a0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:62 +0x543
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildJoin(0xc427cb5960, 0xc428628200, 0xc4285ffc00, 0xc4204d1080)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/join.go:34 +0x73
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildDataSource(0xc427cb5960, 0x68a3140, 0xc428628200, 0x0, 0xc4285ffc00, 0xc420b16780)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:62 +0x543
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildFrom(0xc427cb5960, 0xc428620870, 0xc4285ddbc0, 0xc4285ffc00, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:478 +0xea
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelectClause(0xc427cb5960, 0xc42861f5f0, 0x0, 0x0, 0x0, 0xc4285ffc00, 0x30)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:429 +0x56
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelect(0xc427cb5960, 0xc4286282c0, 0xc4285ffc00, 0x62d0540)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:387 +0x1a6
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildStmt(0xc427cb5960, 0x68c74c0, 0xc4286282c0, 0xc4285ffc00, 0x4)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:177 +0xf1
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build(0xc427cb5960, 0xc400000000, 0x0, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:135 +0x128
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc425308438, 0x68c3ec0, 0xc428620990, 0x68c74c0, 0xc4286282c0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:378 +0x287
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optionallyUseOptimizer(0xc425308438, 0x68c3ec0, 0xc428620990, 0xc42033b810, 0xd, 0xc42513801b, 0x9, 0x0, 0x1, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/planner.go:534 +0xdb
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc425308000, 0x68c3ec0, 0xc428620990, 0x68c74c0, 0xc4286282c0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:751 +0x192
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc425308000, 0x68c3ec0, 0xc428620990, 0x68c74c0, 0xc4286282c0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:401 +0xb82
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc425308000, 0x68c3ec0, 0xc428620990, 0x68c74c0, 0xc4286282c0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:95 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc425308000, 0x68c3e00, 0xc424f8f400, 0xc4203c7b80, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1079 +0x2186
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc4200e5d40, 0x68c3e00, 0xc424f8f400, 0xc42513801b, 0x9, 0xc42513800d, 0x4, 0x0, 0x0, 0x68a5dc0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:416 +0x1bb
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc4200e5d40, 0x68c3e00, 0xc424f8f400, 0xc4209a4a80, 0x5400, 0x15000, 0xc4200fb800, 0xc4203c7b80, 0xc4203c7b70, 0xc4203acff0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:267 +0x122
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:266 +0xf04
```

--- a/pkg/sql/opt/norm/reject_nulls.go
+++ b/pkg/sql/opt/norm/reject_nulls.go
@@ -41,22 +41,22 @@ func (c *CustomFuncs) HasNullRejectingFilter(filter memo.GroupID, nullRejectCols
 }
 
 // NullRejectAggVar scans through the list of aggregate functions and returns
-// the Variable input of the first aggregate that is not ConstAgg or
-// ConstNotNullAgg. Such an aggregate must exist, since this is only called if
-// at least one eligible null-rejection column was identified by the
-// deriveGroupByRejectNullCols method (see its comment for more details).
+// the Variable input of the first aggregate that is not ConstAgg. Such an
+// aggregate must exist, since this is only called if at least one eligible
+// null-rejection column was identified by the deriveGroupByRejectNullCols
+// method (see its comment for more details).
 func (c *CustomFuncs) NullRejectAggVar(aggs memo.GroupID) memo.GroupID {
 	aggsExpr := c.f.mem.NormExpr(aggs).AsAggregations()
 	aggsElems := c.f.mem.LookupList(aggsExpr.Aggs())
 
 	for i := len(aggsElems) - 1; i >= 0; i-- {
 		agg := c.f.mem.NormExpr(aggsElems[i])
-		if agg.Operator() != opt.ConstAggOp && agg.Operator() != opt.ConstNotNullAggOp {
+		if agg.Operator() != opt.ConstAggOp {
 			// Return the input Variable operator.
 			return agg.ChildGroup(c.f.mem, 0)
 		}
 	}
-	panic("couldn't find an aggregate that is not ConstAgg/ConstNotNullAgg")
+	panic("couldn't find an aggregate that is not ConstAgg")
 }
 
 // DeriveRejectNullCols returns the set of columns that are candidates for NULL

--- a/pkg/sql/opt/norm/testdata/rules/reject_nulls
+++ b/pkg/sql/opt/norm/testdata/rules/reject_nulls
@@ -462,3 +462,50 @@ project
       │              └── variable: x [type=int, outer=(5)]
       └── filters [type=bool, outer=(7), constraints=(/7: [/1 - /1]; tight), fd=()-->(7)]
            └── column7 = 1 [type=bool, outer=(7), constraints=(/7: [/1 - /1]; tight)]
+
+# ConstNotNullAgg rejects nulls (regression test for #28810).
+norm
+SELECT 1 FROM a AS ref_0 LEFT JOIN a AS ref_1 ON EXISTS(SELECT 1 FROM a WHERE a.s = ref_0.s)
+----
+project
+ ├── columns: "?column?":17(int!null)
+ ├── fd: ()-->(17)
+ ├── left-join-apply
+ │    ├── columns: a.s:4(string) a.k:5(int) true_agg:15(bool)
+ │    ├── scan a
+ │    │    └── columns: a.s:4(string)
+ │    ├── select
+ │    │    ├── columns: a.k:5(int!null) true_agg:15(bool!null)
+ │    │    ├── outer: (4)
+ │    │    ├── key: (5)
+ │    │    ├── fd: (5)-->(15)
+ │    │    ├── group-by
+ │    │    │    ├── columns: a.k:5(int!null) true_agg:15(bool)
+ │    │    │    ├── grouping columns: a.k:5(int!null)
+ │    │    │    ├── outer: (4)
+ │    │    │    ├── key: (5)
+ │    │    │    ├── fd: (5)-->(15)
+ │    │    │    ├── inner-join
+ │    │    │    │    ├── columns: a.k:5(int!null) a.s:12(string!null) true:14(bool!null)
+ │    │    │    │    ├── outer: (4)
+ │    │    │    │    ├── fd: ()-->(12,14)
+ │    │    │    │    ├── scan a
+ │    │    │    │    │    ├── columns: a.k:5(int!null)
+ │    │    │    │    │    └── key: (5)
+ │    │    │    │    ├── project
+ │    │    │    │    │    ├── columns: true:14(bool!null) a.s:12(string)
+ │    │    │    │    │    ├── fd: ()-->(14)
+ │    │    │    │    │    ├── scan a
+ │    │    │    │    │    │    └── columns: a.s:12(string)
+ │    │    │    │    │    └── projections [outer=(12)]
+ │    │    │    │    │         └── true [type=bool]
+ │    │    │    │    └── filters [type=bool, outer=(4,12), constraints=(/4: (/NULL - ]; /12: (/NULL - ]), fd=(4)==(12), (12)==(4)]
+ │    │    │    │         └── a.s = a.s [type=bool, outer=(4,12), constraints=(/4: (/NULL - ]; /12: (/NULL - ])]
+ │    │    │    └── aggregations [outer=(14)]
+ │    │    │         └── const-not-null-agg [type=bool, outer=(14)]
+ │    │    │              └── variable: true [type=bool, outer=(14)]
+ │    │    └── filters [type=bool, outer=(15), constraints=(/15: (/NULL - ]; tight)]
+ │    │         └── true_agg IS NOT NULL [type=bool, outer=(15), constraints=(/15: (/NULL - ]; tight)]
+ │    └── true [type=bool]
+ └── projections
+      └── const: 1 [type=int]

